R/sheets_speedread.R

```diff
@@ -0,0 +1,221 @@
+#' datasource protocol"
+#'   * <https://stackoverflow.com/a/33727897/2825349>
+#'
+#' The most important link is about "Query Source Ranges":
+#'   * <https://developers.google.com/chart/interactive/docs/spreadsheets#queryurlformat>
+#'
+#' Examples of datasource URLs
+#' ```
+#' # here is the base URL:
+#' https://docs.google.com/spreadsheets/d/SPREADSHEET_ID/gviz/tq
+#'
+#' # examples of the query part
+#' ?tqx=out:csv&sheet={sheet_name}
+#' ?range=A1:C4
+#' ?headers=N
+#' ?gid=N
+#' ````
+#'
+#' A readr issue that demonstrates sending an Authorization header
+#'   * <https://github.com/tidyverse/readr/issues/935>
+#'
+#' @inheritParams sheets_cells
+#' @param .url Whether to use the `"export"` (default) or `"datasource"` URL.
+#' @param .method Whether to process the CSV data with `vroom::vroom()`
+#'   (default) or `readr::read_csv()`.
+#' @param ... Passed along to `vroom::vroom()` or `readr::read_csv()`.
+#'
+#' @return A [tibble][tibble::tibble-package]
+#' @export
+#'
+#' @examples
+#' # Using the large NBA sheet from
+#' # https://github.com/tidyverse/googlesheets4/issues/122
+#' spreadsheet_id <- "1mnWcn7bd7obaXd05rnXrEtgzMBLdy7ctsYvlQM52W00"
+#' (ss <- as_sheets_id(spreadsheet_id))
+#'
+#' sheets_speedread(ss, .url = "export", .method = "vroom")
+#' sheets_speedread(ss, .url = "export", .method = "readr")
+#' sheets_speedread(ss, .url = "datasource", .method = "vroom")
+#' sheets_speedread(ss, .url = "datasource", .method = "readr")
+#'
+#' if (FALSE) {
+#' library(bench)
+#' bnch <- bench::mark(
+#'   sheets_speedread(ss, .url = "export", .method = "vroom"),
+#'   sheets_speedread(ss, .url = "export", .method = "readr"),
+#'   sheets_speedread(ss, .url = "datasource", .method = "vroom"),
+#'   sheets_speedread(ss, .url = "datasource", .method = "readr"),
+#'   iterations = 5,
+#'   check = FALSE
+#' )
+#' bnch
+#' }
+#'
+#' # prove that we can send a cell range, sheet, and col spec through ...
+#' (df <- sheets_speedread(
+#'   sheets_example("deaths"),
+#'   sheet = "other",
+#'   range = "A5:F15",
+#'   col_types = vroom::cols(
+#'     Age = vroom::col_integer(),
+#'     `Date of birth` = vroom::col_date("%m/%d/%Y"),
+#'     `Date of death` = vroom::col_date("%m/%d/%Y")
+#'   ),
+#'   .method = "vroom"
+#' ))
+#'
+#'
+#' if (FALSE) {
+#' # write a private Sheet
+#' (ss <- sheets_write(iris))
+#'
+#' # read via 'export' URL (sends a bearer token)
+#' sheets_speedread(ss, .url = "export")
+#'
+#' # read via 'datasource' URL (access token goes as query param)
+#' sheets_speedread(ss, .url = "datasource")
+#'
+#' # clean up
+#' googledrive::drive_trash(ss)
+#' }
+sheets_speedread <- function(ss,
+                             sheet = NULL,
+                             range = NULL,
+                             skip = 0,
+                             ...,
+                             .url = c("export", "datasource"),
+                             .method = c("vroom", "readr")) {
```

Comment:
Again, not really sure if I should support both URL forms and both packages, but it was worth studying.